File: python/hbb1D.py
#!/usr/bin/env python3
from partitioning1D import Partitioning1D
from partitioning import IndexResult


all_hbb1ds = dict()

saved = 0
def HBB1D_factory(geom_info, halo, dimension):
    key = (geom_info, halo, dimension)
    if key not in all_hbb1ds:
        all_hbb1ds[key] = _HBB1D(*key)
    else:
        global saved
        saved += 1
    return all_hbb1ds[key]


class _HBB1D(Partitioning1D):
    def __init__(self, geom_info, halo, dimension):
        self.size, self.parity = geom_info
        self.dimension = dimension
        self.halo = halo

        assert halo != 0
        assert self.size > 2 * halo

    def _key(self):
        return (self.size,self.parity,self.dimension,self.halo)

    @property
    def limits(self):
        return [-self.halo, 0, self.halo, self.size - self.halo, self.size, self.size + self.halo]

    @property
    def comments(self):
        return f" HBB1D - Dimension: {self.dimension}, size: {self.size}, parity: {self.parity}, halo: {self.halo}"

    def coord_to_idxs(self, relative_x):
        if not self.limits[0] <= relative_x < self.limits[-1]:
            return []
        else:
            idx, local_rest = next(
                (i, relative_x - s)  #
                for i, (s, e) in enumerate(zip(self.starts, self.ends))
                if s <= relative_x < e)

            return [
                IndexResult(
                    idx=idx,  #
                    rest=local_rest,  #
                    cached_flag=False)
            ]

    def idx_to_coord(self, idx, offset):
        return self.starts[idx] + offset

    def max_idx_value(self):
        return 5

File: cpp/src/1D/hbb1D.cpp
#include "1D/hbb1D.hpp"
#include <cassert>

namespace hypercubes {
namespace slow {
HBB1D::HBB1D(SizeParity sp, int dimension_, std::string name_, int halo_)
    : Partitioning1D(sp, dimension_, name_), halo(halo_) {
  assert(halo != 0);
  assert(size > 2 * halo);
  limits = std::vector<int>{-halo, 0, halo, size - halo, size, size + halo};
}
int HBB1D::idx_to_coord(int idx, int offset) const {
  return start(idx) + offset;
}
int HBB1D::max_idx_value() const { return 5; };
std::string HBB1D::comments() const { return tuple_to_str(key()); };
std::vector<IndexResult> HBB1D::coord_to_idxs(int relative_x) const {
  for (int i = 0; i < 5; ++i)
    if (start(i) <= relative_x and relative_x < end(i)) {
      return std::vector<IndexResult>{{i, relative_x - start(i), false}};
    }
  return std::vector<IndexResult>();
}
} // namespace slow
} // namespace hypercubes
